### PR TITLE
Added new methods and rewrited constructor in math_vec3.js

### PR DESCRIPTION
**Pull request by eXponenta**
Add new methods:
- reflect - It reflects one vector from the normal. Ex: (-1,1,0).reflect(0,1,0) = (-1,1,0)
- maxLength,minLength - return maximum and minimum between the two vectors of length 
- toJSON - return JSON data, see JSON.stringify()
- get - return array of vector data

Rewrited constructor. Now it takes array[3]/[2]/[1] and from 1 to 3 arguments as number

### CODE REVIEW

**Comment by guycalledfrank** [comment]
I'm not sure it's obvious that min/max should compare by length. More often vec3 min/max implementations are meant for per-component comparison.

Also, reflect should have version without allocations (clone = memory allocation), that would use supplied vec3 object as output.

File: src/math/math_vec3.js [hunk 1]
```diff
@@ -457,6 +467,38 @@ pc.extend(pc, (function () {
 
             return this;
         },
+        //reflect vector of normal
+        //ex: (-1,1,0).reflect(0,1,0) -> (-1,1,0)
+        
+	    reflect: function(n){
+            var _n = n.clone();
```
Comment (eXponenta) [inline src/math/math_vec3.js]: Don't use CLONE

**Comment by eXponenta** [comment]
Fixed!

**Comment by Maksims** [comment]
toString is the same as toJSON. And toJSON probably should return JSON, which is actually an object, not stringified version. Pretty much: `this.get();`

**Comment by eXponenta** [comment]
Oh, Thanks! Fixed

**Comment by willeastcott** [comment]
Can we slow down a bit here. It's arguable to add reflect. But minLength and maxLength are just very simple utility functions. I'm not convinced they're adding much value.

    var min = v1.length() < v2.length ? v1 : v2;

What's wrong with just having the above in your code?

It's also arguable to add toJSON perhaps, but wouldn't it be:

return { x: this.x, y: this.y, z: this.z };

? Also, we can't just add a generic function like that without also doing the other math classes.

As for the constructor, we could have it take any combination of parameters (numbers, arrays, typed arrays, strings, and so on), but there's a balance between maintaining simplicity, speed and flexibility. Plus, any change to one math class constructor really ought to apply to all of them.

My current feeling is I'm OK to pull reflect. With documentation included. :)

**Comment by Maksims** [comment]
If found lack of constructor from array pretty often.
It is sad that we do `prototype = { ...` instead of individual method, as `Vec3.prototype.toJSON = Vec3.prototype.get` - would be more optimal and efficient.

**Comment by eXponenta** [comment]
I can add reflect method to 'math_vec2.js' , if it's necessary.
I'm think that creating vectors from array too necessary, because it's convenient when used JSON. 
For example:
```javascript
var pos  = new pc.Vec3(p); // p - array 
```
instead of

```javascript
var pos  = new pc.Vec3(p[0],p[1],p[2]||0);
```

**Comment by Maksims** [comment]
Closing this PR due to lack of activity on our end, and since then many things have changed regarding Vec3. Although `reflect` method could be introduced as separate PR in the future and is considered useful.

**Pull request closed without merging**